### PR TITLE
Add focal_loss

### PR DESCRIPTION
**Pull request by Kongsea**
Add a new loss function, focal_loss, which was described in [this paper](https://arxiv.org/pdf/1708.02002.pdf).

### CODE REVIEW

**Comment by fchollet** [comment]
This PR adds a new loss in core TF, `focal_loss`, which was described in an unreviewed paper that was released on Arxiv on August 7, two weeks ago.

I believe we should not add to core TF (in near-real time) every new idea that gets pushed to Arxiv. All code in core TF needs to be maintained in the long-term: adding new code has a cost, which gets integrated over time. Until it is demonstrated that a given feature is useful, we should not add it to core TF.

**Comment by martinwicke** [comment]
Can you add this to contrib (contrib/losses) instead?

**Comment by Kongsea** [comment]
OK, I will move it to contrib.

**Comment by mrlzla** [comment]
Is this function working correct?
I have constant loss all the train time.

**Comment by drpngx** [comment]
@Kongsea any luck with the move to contrib?

**Comment by Kongsea** [comment]
@drpngx I have moved it to contrib. Please check it. Thank you.

**Comment by drpngx** [comment]
There has been a change, I think you need to pull rebase and push again. Thanks.

**Comment by Kongsea** [comment]
Now it seems OK. Please check it again.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by CasiaFan** [comment]
Based on the definition of Pt in paper, should the `labels` be added to this [equation](https://github.com/tensorflow/tensorflow/pull/12257/files#diff-01ec5a7b62cbe21e7fda63c7a92947e6R716)? @Kongsea

**Comment by CasiaFan** [comment]
Based on the definition of cross-entropy, should the `labels` be added to this [equation](https://github.com/tensorflow/tensorflow/pull/12257/files#diff-01ec5a7b62cbe21e7fda63c7a92947e6R716)? otherwise, when alpha=0, gamma=1, the FL value is different from what tf.nn.softmax_cross_entropy returns.

**Comment by sb2nov** [comment]
@vrv please review

**Comment by sb2nov** [comment]
Jenkins, test this please.

**Comment by vrv** [comment]
Adding gpapan@ who knows about this area, or feel free to re-assign it to someone who knows it well (I don't know it well enough).

**Comment by martinwicke** [comment]
@gpapan are you able to review this? Let me know if not.

**Comment by martinwicke** [comment]
CLA confused by me.

**Comment by drpngx** [comment]
@martinwicke care to suggest an alternative reviewer perhaps? It looks like @gpapan is not available.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

File: tensorflow/contrib/losses/python/losses/loss_ops.py [hunk 1]
```diff
@@ -664,3 +665,60 @@ def cosine_distance(
     radial_diffs = math_ops.multiply(predictions, labels)
     losses = 1 - math_ops.reduce_sum(radial_diffs, reduction_indices=[axis,])
     return compute_weighted_loss(losses, weights, scope=scope)
+
+
+def focal_loss(predictions, labels, gamma=2, alpha=1, weights=1.0,
```
Comment (alextp) [inline tensorflow/contrib/losses/python/losses/loss_ops.py]: The arguments to losses should be labels,predictions instead of predictions,labels

File: tensorflow/contrib/losses/python/losses/loss_ops.py [hunk 2]
```diff
@@ -664,3 +665,60 @@ def cosine_distance(
...
+  `weights` acts as a coefficient for the loss. If a scalar is provided, then
+  the loss is simply scaled by the given value. If `weights` is a tensor of size
+  [batch_size], then the total loss for each sample of the batch is rescaled
+  by the corresponding element in the `weights` vector. If the shape of
+  `weights` matches the shape of `predictions`, then the loss of each
+  measurable element of `predictions` is scaled by the corresponding value of
+  `weights`.
+
+  Args:
+    labels: The ground truth output tensor, same dimensions as 'predictions'.
```
Comment (alextp) [inline tensorflow/contrib/losses/python/losses/loss_ops.py]: Should labels be [0, 1], [-1, 1]? Floating points? Integers?

**Review by alextp — CHANGES_REQUESTED** [review]

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by bhack** [comment]
Any news?

**Comment by bhack** [comment]
@tensorflowbutler Can you label nagging issue as nagging?
